src/dropdown_select.rs

```diff
@@ -0,0 +1,333 @@
+                        .into_iter()
+                        .find_map(|(label, variant)| if *t == variant { Some(label) } else { None })
+                        .unwrap()
+                })
+                .on_click(|ctx: &mut EventCtx, t: &mut DropdownState<T>, _| {
+                    t.expanded = true;
+                    ctx.submit_notification(DROP)
+                }),
+                move |_t: &DropdownState<T>, env: &Env| {
+                    ControllerHost::new(
+                        ListSelect::new(values.clone())
+                            .lens(DropdownState::<T>::data)
+                            .border(env.get(theme::BORDER_DARK), 1.0),
+                        DropdownSelect { _t: PhantomData },
+                    )
+                },
+            ),
+        )
+    }
+}
+
+// This controller will send itself "COLLAPSE" events whenever the dropdown is removed, and
+// reacts to it by updating its expanded state
+impl<T: Data + PartialEq> Controller<DropdownState<T>, Container<DropdownState<T>>>
+    for DropdownSelect<T>
+{
+    fn event(
+        &mut self,
+        child: &mut Container<DropdownState<T>>,
+        ctx: &mut EventCtx,
+        event: &Event,
+        data: &mut DropdownState<T>,
+        env: &Env,
+    ) {
+        match event {
+            Event::Command(n) if n.is(COLLAPSE) => {
+                data.expanded = false;
+            }
+            _ => child.event(ctx, event, data, env),
+        }
+    }
+    fn lifecycle(
+        &mut self,
+        child: &mut Container<DropdownState<T>>,
+        ctx: &mut LifeCycleCtx,
+        event: &LifeCycle,
+        data: &DropdownState<T>,
+        env: &Env,
+    ) {
+        if let LifeCycle::HotChanged(false) = event {
+            ctx.submit_command(COLLAPSE);
+        }
+        child.lifecycle(ctx, event, data, env)
+    }
+
+    fn update(
+        &mut self,
+        child: &mut Container<DropdownState<T>>,
+        ctx: &mut UpdateCtx,
+        old_data: &DropdownState<T>,
+        data: &DropdownState<T>,
+        env: &Env,
+    ) {
+        ctx.submit_command(COLLAPSE);
+        child.update(ctx, old_data, data, env)
+    }
+}
+
+#[derive(Clone, Data, Lens)]
+struct DropdownState<T> {
+    data: T,
+    has_children: bool,
+    expanded: bool,
+}
+
+impl<T> DropdownState<T> {
+    fn new(data: T) -> Self {
+        DropdownState {
+            data,
+            has_children: true,
+            expanded: false,
+        }
+    }
+}
+
+/// A button with a left or down arrow, changing shape when opened.
+struct DropdownButton<T> {
+    wedge: WidgetPod<(bool, bool), Wedge>,
+    label: Label<T>,
+    label_size: Size,
+}
+
+impl<T: Data> DropdownButton<T> {
+    fn new(text: impl Into<LabelText<T>>) -> DropdownButton<T> {
+        DropdownButton::from_label(Label::new(text))
+    }
+
+    fn from_label(label: Label<T>) -> DropdownButton<T> {
+        DropdownButton {
+            wedge: WidgetPod::new(Wedge::new()),
```

Comment:
It feels like it would make sense to lens this to the tuple you want rather than manually doing it in each method?
Ah, this one would make it easier https://github.com/linebender/druid/pull/1540 ...

Reply:
Yes, but unfortunately it's not merged yet... 
I've chosen to simplify the `Wedge` widget to have it work with boolean, this removes the (somewhat ugly) tuple.

If I was to lens it, the concrete `Wedge` type in the definition of the struct `DropdownButton` would get replaced by something like `LensWarp<bool, DropdownState<T>, Lens<bool, DropdownState<T>>, Wedge>`, that's quite long, is there a better choice?

Reply:
I don't know if there is a better way at the moment really. It is either that or boxing I think.
I do sometimes put in type aliases for these composed things, which can at least make the struct look less noisy